### DISCUSSION TITLE
meaning of the values for hyperprior

### DISCUSSION THREAD

**Discussion by AlbertoPresta** [InterDigitalInc/CompressAI#170] [Q&A]
Hi,
sorry if I bother you again, but I have a question related to these variables (I am talking about the hyperprior model);

SCALES_MIN = 0.11
SCALES_MAX = 256
SCALES_LEVELS = 64

How do you use these variable in the update() function? 

Alberto

**Comment by YodaEmbedding** [comment]
"Scale" refers to the "width" or "standard deviation" of the normal probability distribution.

https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/models/google.py#L196-L203

`get_scale_table()` is used to create 64 standard deviations ranging from σ = 0.11 to σ = 256. The standard deviations are listed as follows:

> [0.11, 0.12, 0.14, 0.16, 0.18, 0.20, 0.23, 0.26, 0.29, 0.33, 0.38, 0.43, 0.48, 0.54, 0.62, 0.70, 0.79, 0.89, 1.01, 1.14, 1.29, 1.46, 1.65, 1.86, 2.11, 2.38, 2.70, 3.05, 3.45, 3.90, 4.41, 4.99, 5.64, 6.38, 7.22, 8.16, 9.23, 10.44, 11.81, 13.35, 15.10, 17.08, 19.32, 21.85, 24.71, 27.94, 31.60, 35.74, 40.42, 45.71, 51.70, 58.47, 66.13, 74.79, 84.58, 95.65, 108.18, 122.35, 138.37, 156.48, 176.98, 200.15, 226.36, 256.00]

The following then uses each standard deviation to define a Gaussian distribution (i.e. `_standardized_cumulative`):

https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/entropy_models/entropy_models.py#L625-L640

**Reply by AlbertoPresta** [reply]
ok thanks! maybe is a stupid question, but in the end the quantized_cdf is referred to what? to the guassian latent space? I do not get the reasons why yiu computed the cdf on the domain extracted from the scales. In the end should we have a standard deviation for each dimension of the latent space?

**Reply by AlbertoPresta** [reply]
And how it is used then the quantized_cdf in the compress method? it should represent the distribution of the latent space, but I do not get the meaning of the variables indexes also.

**Reply by YodaEmbedding** [reply]
- **For all `EntropyModel`s:** The `symbols` are the actual tensor values that we want to encode. The `indexes` represent the CDFs used to encode various symbols.
 - **For `EntropyBottleneck`**: `_quantized_cdf` is a table of non-parametric distributions that depend on the `matrix`es and `bias`es. The CDFs generated by `update()` thus change after a training epoch. In the CompressAI implementation, the `indexes` are `[0, 0, ..., 1, 1, ..., 2, 2, ..., 191, 191]` since an entire channel is encoded using only a single corresponding CDF.
 - **For `GaussianConditional`:** `_quantized_cdf` is a table of various pregenerated Gaussian CDFs whose standard deviations range over `[0.11, 256.00]`. The `indexes` allow us to choose which Gaussian CDF we are talking about. Thus, there is a one-to-one mapping between `indexes` and standard deviations. In fact, we compute the `indexes` directly using the model-predicted `scales` when we call the `build_indexes` function.
